### ISSUE TITLE
[feature request] ABT_pool_create_unit() and ABT_pool_free_unit()

### ISSUE THREAD

**Issue by mdorier** [pmodels/argobots#378]
I am trying to write an `ABT_pool_user_def` pool definition for a pool that redirects its calls to another `ABT_pool` (kind of a "wrapper" pool). I have multiple use-cases for that (including a pool that logs the timings of ULT it sees being pushed and popped). The problem is, I can't wrap all the functionalities of the inner pool without having access to more information about it.

Example:

```cpp
/* Function I can wrap with the currently provided API */
ABT_bool pool_wrapper_is_empty_fn(ABT_pool p)
{
     ABT_pool inner = ... /* get inner pool using ABT_pool_get_data on p */
     ABT_bool result;
     ABT_pool_is_empty(inner, &result);
     return result;
}

/* Function I cannot wrap with the currently provided API */
ABT_unit pool_wrapper_create_unit_fn(ABT_pool p, ABT_thread t)
{
    ABT_pool inner = ... /* get inner pool using ABT_pool_get_data on p */
    /* there is no ABT_pool_create_unit() */
}
```

I think all I need is for Argobots to provide public `ABT_pool_create_unit()` and `ABT_pool_free_unit()` that would call the underlying pool definition's function pointers.

Alternatively, this could be achieved with an `ABT_pool_get_user_def()` that would give access to the pool's `ABT_pool_user_def`, and some `ABT_pool_user_def_get_*()` function to retrieve the function pointers from the pool definition.

**Comment by mdorier** [comment]
Actually thinking a bit more about it, I think I could just have `ABT_unit = ABT_thread` in the context of the wrapper pool, since I don't have to attach any information to the thread. My `pool_wrapper_create_unit_fn` would simply do a cast, and my `pool_wrapper_free_unit_fn` would be empty...

**Comment by shintaro-iwasaki** [comment]
This is a very good point. I don't think nested `ABT_pool` (or a `ABT_pool` wrapper using `ABT_pool_user_def`) should be allowed within the current Argobots mechanism because `ABT_unit` (or `ABT_thread`) must be associated with only one `ABT_pool` ... as far as I remember.

1. `ABT_pool_push()` that pushes `thread1` to `pool1`, for example, internally changes `thread1`'s associated pool to `pool1` and call its user-defined `p_pool_push()`.
2. The user-defined `p_pool_push()` internally calls `ABT_pool_push()` that pushes `thread1` to a child pool, say `pool2`. Internally, `thread1`'s associated pool will be changed from `pool1` to `pool12 
3. When `ABT_pool_pop()` tries to pop `thread1` from `pool1`, `thread1` is associated with `pool2`. This can cause some issues.

In reality, maybe you don't see an issue particularly if you don't enable extra assertion (e.g., undefined-behavior assert etc), but I wouldn't recommend it.

Note: fixing this would need pool management change, so allowing nested pools might not be trivial

**Comment by shintaro-iwasaki** [comment]
> What I don't understand yet is the role of the `create_unit` and `free_unit` functions, and when exactly they are called. Do I understand correctly that the following happens:

Yes mostly correct.  Argobots creates and frees units lazily, so they are called when ULT's associated pool is changed (for example, by `ABT_thread_set_associated_pool`), not when a ULT is popped ... as far as I remember.

> EDIT 2: if the child pool is a builtin pool, then it looks like there won't be any allocation/deallocation:

Yes this is part of the optimizations.  Basically `ABT_unit == ABT_thread` (though there exist some bit tricks as you see) for built-in pool implementations.

> Here is my full code example, let me know if you see a scenario where it could fail

I cannot come up with exactly when, but more specifically:
- It is undefined which pool `ABT_thread_get_last_pool()` or `ABT_self_get_last_pool()` would return (i.e., `wrapper_pool` or `data->pool`).
- Similarly to the first case, it is undefined to which pool a ULT is pushed when `ABT_self_yield()` is called.

It'd be okay if it is working in your specific use case, but I would not highly recommend it since it's undefined.

> How can I enable the assertions you were talking about?

To enable an UB assert, please use `./configure --enable-debug=most`(see #335) but maybe this does not flag nested pools.

**Comment by mdorier** [comment]
I'm looking at the code, and I'm under the impression that those functions (`ABT_thread_get_last_pool`, `ABT_self_get_last_pool`, and `ABT_self_yield`) would actually have a well defined behavior. The key is that I'm calling `ABT_thread_set_associated_pool` on the threads on their way in and out of the pool to maintain a correct information of which pool the thread is in. But maybe I'm missing something...

Note also: I simplified a bit my code above but in practice my wrapper pool would have a mutex protecting every operation, so that I ensure correct behavior in an MPMC setting).

Definition of `ABTI_thread`:
https://github.com/pmodels/argobots/blob/915ff282e2ace23eeb7d9147c03b7ebc4fd9d5e7/src/include/abti.h#L422-L437

Definition of `ABT_thread_get_last_pool()`:
https://github.com/pmodels/argobots/blob/915ff282e2ace23eeb7d9147c03b7ebc4fd9d5e7/src/thread.c#L1109-L1119

Definition of `ABT_self_get_last_pool()`:
https://github.com/pmodels/argobots/blob/915ff282e2ace23eeb7d9147c03b7ebc4fd9d5e7/src/self.c#L468-L479

Definition of `ABT_self_yield()`:
https://github.com/pmodels/argobots/blob/915ff282e2ace23eeb7d9147c03b7ebc4fd9d5e7/src/self.c#L626-L638

The `ABT_self_*` functions have to be called by the ULT when the ULT is running, i.e. when it's not in the pool, and because of the call to `ABT_thread_set_associated_pool` in my wrapper pool's `pop` function, the `p_pool` pointer in the `ABTI_thread` will necessarily point to the wrapper pool.

`ABT_thread_get_last_pool` could be called in any context, and the result would depend on whether the thread is running (in which case the wrapper pool will be returned) or not (in which case it will be inside the child pool and this is what the function will return). This seems like a well-defined behavior.

I think the only way I could get an undefined behavior is if someone calls `ABT_thread_get_last_pool` at the exact same time my wrapper calls `ABT_thread_set_associated_pool`, more specifically at this exact line: https://github.com/pmodels/argobots/blob/915ff282e2ace23eeb7d9147c03b7ebc4fd9d5e7/src/include/abti_unit.h#L209 because it's not an atomic assignment. That said, this begs the question: when is `ABT_thread_set_associated_pool` actually safe to call? Because a custom scheduler, or any ULT, could just as well call this function to change the pool a thread is associated with.

**Comment by shintaro-iwasaki** [comment]
Sorry for the late reply. Maybe I used a word "undefined behavior" in a confusing way:

- From the runtime developer's viewpoint, they are all classified into undefined behavior since API does not specify it and its behavior cannot be naturally deduced. Since it's undefined, the runtime might not work as intended (since it's unclear what can be natural intention). At least, I didn't assume this nested pool situation when I developed Argobots, so anything can happen. This is why I don't recommend it.
   - My "undefined behavior" means:
     - The API does not define which pool `ABT_thread_get_last_pool()` or `ABT_self_get_last_pool()` would return (i.e., wrapper_pool or `data->pool`).
     - The API does not define which pool a ULT is pushed when `ABT_self_yield()` is called.

- If your "well-defined behavior" means "Argobots works without SEGV", perhaps it works if the user doesn't call `ABT_thread_set_associated_pool()` and the user doesn't care about to which pool a ULT is pushed when `ABT_self_yield()` is called. Argobots neither tests nor supports this behavior, so please test if it works in your workload.

> when is `ABT_thread_set_associated_pool` actually safe to call? Because a custom scheduler, or any ULT, could just as well call this function to change the pool a thread is associated with.

`ABT_thread_set_associated_pool()` should be called when a ULT is "not" in a pool. Since a ULT is not in a pool, this ULT's pool won't be changed by another scheduler.

```cxx
// pseudo code, typical use case
thread = ABT_pool_pop(pool1);
// If this thread yields, this thread will be pushed to pool2, not pool1.
ABT_thread_set_associated_pool(thread, pool2);
ABT_self_schedule(thread);
```

**Comment by mdorier** [comment]
Oh I see, the problem is that an Argobots developer could change something in the implementation that would change the behavior, because that behavior is undocumented. So my code works fine with the current implementation of Argobots but may not in the future.
